File: docs/examples/webhook-receiver.go
// A complete, runnable webhook receiver for the SIP platform (no dependencies).
//
// Verifies the HMAC-SHA256 signature on every delivery, then dispatches events.
// Set WEBHOOK_SECRET to the signing secret from the portal (Tenant -> Webhooks)
// and point your endpoint URL at this server (expose it over HTTPS, e.g. via a
// tunnel — the platform only delivers to public HTTPS URLs).
//
//	WEBHOOK_SECRET=... go run webhook-receiver.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
)

func main() {
	secret := os.Getenv("WEBHOOK_SECRET")
	if secret == "" {
		log.Fatal("set WEBHOOK_SECRET to your endpoint's signing secret")
	}
	http.HandleFunc("/webhook", func(w http.ResponseWriter, r *http.Request) {
		raw, _ := io.ReadAll(r.Body) // the EXACT bytes — verify over these
		mac := hmac.New(sha256.New, []byte(secret))
		mac.Write(raw)
		expected := "sha256=" + hex.EncodeToString(mac.Sum(nil))
		if !hmac.Equal([]byte(expected), []byte(r.Header.Get("X-Webhook-Signature"))) {
			http.Error(w, "bad signature", http.StatusUnauthorized)
			return
		}

		var env struct {
			Event string         `json:"event"`
			Data  map[string]any `json:"data"`
		}
		_ = json.Unmarshal(raw, &env)

		switch env.Event { // call.completed, trunk.down/up, voicemail.new, test.ping
		case "call.completed":
			fmt.Printf("call %v %v -> %v %vs %v\n",
				env.Data["direction"], env.Data["from"], env.Data["to"],
				env.Data["duration_sec"], env.Data["disposition"])
		case "trunk.down", "trunk.up":
			fmt.Printf("trunk %v %v -> %v\n", env.Data["trunk"], env.Data["prev_state"], env.Data["state"])
		case "voicemail.new":
			fmt.Printf("new voicemail for ext %v from %v\n", env.Data["extension_id"], env.Data["caller_id_num"])
		case "test.ping":
			fmt.Println("test ping received — endpoint is wired up correctly")
		default:
			fmt.Printf("unhandled event: %s\n", env.Event)
		}

		w.WriteHeader(http.StatusNoContent) // return 2xx fast; non-2xx triggers retry
	})

	addr := ":8088"
	log.Printf("listening on %s/webhook", addr)
	log.Fatal(http.ListenAndServe(addr, nil))
}

File: docs/examples/webhook-receiver.py
#!/usr/bin/env python3
"""
A complete, runnable webhook receiver for the SIP platform.

It verifies the HMAC-SHA256 signature on every delivery (rejecting forgeries)
and dispatches the events the platform sends. Copy it, set WEBHOOK_SECRET to the
signing secret shown when you create the endpoint in the portal
(Tenant -> Webhooks), and point your endpoint URL at this server.

  pip install flask           # only dependency
  WEBHOOK_SECRET=... python3 webhook-receiver.py
  # expose it publicly (e.g. `ngrok http 8088`) and register that HTTPS URL.

Only public HTTPS URLs are accepted by the platform, and redirects are not
followed — terminate TLS in front of this (a tunnel like ngrok does that).
"""
import hashlib
import hmac
import os
import sys

from flask import Flask, request

SECRET = os.environ.get("WEBHOOK_SECRET", "")
if not SECRET:
    sys.exit("set WEBHOOK_SECRET to your endpoint's signing secret")

app = Flask(__name__)


def signature_ok(raw_body: bytes, header: str) -> bool:
    expected = "sha256=" + hmac.new(SECRET.encode(), raw_body, hashlib.sha256).hexdigest()
    # constant-time compare avoids leaking the secret via timing.
    return hmac.compare_digest(expected, header or "")


@app.post("/webhook")
def webhook():
    raw = request.get_data()  # the EXACT bytes — sign/verify over these, not a re-encode
    if not signature_ok(raw, request.headers.get("X-Webhook-Signature", "")):
        return ("bad signature", 401)

    event = request.headers.get("X-Webhook-Event", "")
    payload = request.get_json(silent=True) or {}
    data = payload.get("data", {})

    # Dispatch. Events: call.completed, trunk.down, trunk.up, voicemail.new, test.ping
    if event == "call.completed":
        print(f"call {data.get('direction')} {data.get('from')} -> {data.get('to')} "
              f"{data.get('duration_sec')}s {data.get('disposition')}")
    elif event in ("trunk.down", "trunk.up"):
        print(f"trunk {data.get('trunk')} {data.get('prev_state')} -> {data.get('state')}")
    elif event == "voicemail.new":
        print(f"new voicemail for ext {data.get('extension_id')} from {data.get('caller_id_num')}")
    elif event == "test.ping":
        print("test ping received — endpoint is wired up correctly")
    else:
        print(f"unhandled event: {event}")

    # Return 2xx quickly; the platform retries on non-2xx / timeout.
    return ("", 204)


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=int(os.environ.get("PORT", "8088")))
